File: src/main/GO/leetcode/lists/hot100/LeetCode438.go
package hot100

func findAnagrams(s string, p string) []int {
	need, window := make(map[byte]int), make(map[byte]int)
	left, right := 0, 0
	valid := 0
	res := make([]int, 0)
	for i := range p {
		need[p[i]] = need[p[i]] + 1
	}
	for right < len(s) {
		c := s[right]
		right++
		//进行窗口内数据的一系列更新
		if _, ok := need[c]; ok {
			window[c] = window[c] + 1
			if need[c] == window[c] {
				valid++
			}
		}
		for right-left >= len(p) {
			if valid == len(need) {
				res = append(res, left)
			}

			d := s[left]
			if _, ok := need[d]; ok {
				if window[d] == need[d] {
					valid--
				}
				window[d] = window[d] - 1
			}
			left++
		}
	}
	return res
}

File: src/main/java/leetcode/lists/hot100/LeetCode438.java
package leetcode.lists.hot100;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/20 02:10
 */
public class LeetCode438 {

    public List<Integer> findAnagrams(String s, String t) {
        Map<Character, Integer> need = new HashMap<>();
        Map<Character, Integer> window = new HashMap<>();
        for (char c : t.toCharArray()) {
            need.put(c, need.getOrDefault(c, 0) + 1);
        }

        int left = 0, right = 0, count = 0;
        List<Integer> list = new ArrayList<>();
        while (right < s.length()) {
            char c = s.charAt(right);
            right++;
            if (need.containsKey(c)) {
                window.put(c, window.getOrDefault(c, 0) + 1);
                if (need.get(c).equals(window.get(c))) {
                    count++;
                }
            }
            //只有在right - left == need.size()的情况下才有可能有符合条件的解
            while (right - left >= t.length()) {
                if (count == need.size()) {
                    list.add(left);
                }
                char d = s.charAt(left);
                if (need.containsKey(d)) {
                    if (need.get(d).equals(window.get(d))) {
                        count--;
                    }
                    window.put(d, window.getOrDefault(d, 0) - 1);
                }
                left++;
            }
        }
        return list;
    }
}
